File: server/src/main/java/bgu/spl/net/impl/stomp/Frame.java
package bgu.spl.net.impl.stomp;

import java.util.Map;
import java.util.HashMap;

public class Frame {
    private String command;
    private Map<String, String> headers = new HashMap<>();
    private String body;

    public Frame(String command) {
        this.command = command;
    }

    public void addHeader(String key, String value) {
        headers.put(key, value);
    }

    public void setBody(String body) {
        this.body = body;
    }

    public String getCommand() {
        return command;
    }

    public Map<String, String> getHeaders() {
        return headers;
    }

    public String getHeader(String key) {
        return headers.get(key);
    }

    public String getBody() {
        return body;
    }

    public static Frame parseFrame(String rawFrame) {
        if (rawFrame == null || rawFrame.isEmpty()) {
            throw new IllegalArgumentException("Frame is empty or null");
        }

        String[] lines = rawFrame.split("\n");
        if (lines.length == 0) {
            throw new IllegalArgumentException("Frame has no content");
        }

        // First line is the command
        String command = lines[0].trim();
        if (command.isEmpty()) {
            throw new IllegalArgumentException("Frame command is empty");
        }

        Frame frame = new Frame(command);

        // Parse headers
        int i = 1;
        while (i < lines.length && !lines[i].trim().isEmpty()) {
            String[] header = lines[i].split(":", 2);
            if (header.length != 2) {
                throw new IllegalArgumentException("Invalid header format: " + lines[i]);
            }
            frame.addHeader(header[0].trim(), header[1].trim());
            i++;
        }

        // Parse body (if any)
        if (i + 1 < lines.length) {
            StringBuilder body = new StringBuilder();
            for (int j = i + 1; j < lines.length; j++) {
                body.append(lines[j]).append("\n");
            }
            frame.setBody(body.toString().trim());
        }

        return frame;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append(command).append("\n");
        headers.forEach((key, value) -> sb.append(key).append(":").append(value).append("\n"));
        if (body != null && !body.isEmpty()) {
            sb.append("\n").append(body);
        }
        sb.append('\u0000');
        return sb.toString();
    }
}

File: client/src/Frame.cpp

#include "../include/Frame.h"
#include <sstream>
#include <stdexcept>

Frame::Frame(const std::string &command) : command(command) {}

// Frame* Frame::clone() const
// {
//     Frame *frame = new Frame(command);
//     frame->headers = headers;
//     frame->body = body;
//     return frame;
// }

Frame::~Frame() {}
void Frame::addHeader(const std::string &key, const std::string &value)
{
    headers[key] = value;
}

void Frame::setBody(const std::string &body)
{
    this->body = body;
}

std::string Frame::getCommand() const
{
    return command;
}

std::map<std::string, std::string> Frame::getHeaders() const
{
    return headers;
}

std::string Frame::getHeader(const std::string &key) const
{
    auto it = headers.find(key);
    return (it != headers.end()) ? it->second : "";
}

std::string Frame::getBody() const
{
    return body;
}

Frame Frame::parseFrame(const std::string &rawFrame)
{
    std::istringstream stream(rawFrame);
    std::string line, command;
    std::getline(stream, command);
    Frame frame(command);

    while (std::getline(stream, line) && !line.empty())
    {
        auto delimiterPos = line.find(':');
        if (delimiterPos != std::string::npos)
        {
            std::string key = line.substr(0, delimiterPos);
            std::string value = line.substr(delimiterPos + 1);
            frame.addHeader(key, value);
        }
    }

    std::string body;
    while (std::getline(stream, line))
        body += line + "\n";

    if (!body.empty())
        frame.setBody(body.substr(0, body.size() - 1));
    return frame;
}

std::string Frame::toString() const
{
    std::ostringstream sb;
    sb << command << "\n";
    for (const auto &header : headers)
        sb << header.first << ":" << header.second << "\n";

    if (!body.empty())
        sb << "\n"
           << body;
    return sb.str();
}

std::string Frame::getUserNameFromBody() const
{
    std::istringstream bodyStream(this->getBody());
    std::string line;
    std::string user;

    while (std::getline(bodyStream, line))
    {
        if (line.find("user:") == 0)
        {
            user = line.substr(5); // Extract the value after "user:"
            break;
        }
    }
    return user;
}